File: Class1Q/ProblemSet/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long getNumber();
int isVisa(long number);
int main(void)
{
    long id = getNumber();
    int flag = isVisa(id);
    if (flag == 1)
    {
        printf("VISA\n");
    }
    else if (flag == 2)
    {
        printf("MASTERCARD\n");
    }
    else if (flag == 3)
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;

}
long getNumber()
{
    long number = 0;
    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);
    return number;
}
int isVisa(long number)
{
    // 计算*2后的单个数字和
    long sum = 0;
    // 计算位数
    long digit = 0;
    // 取后两位
    long tail = 0;

    long temp = number;
    while (temp > 0)
    {
        tail = temp % 100;
        tail /= 10;
        tail *= 2;
        if (tail >= 10)
        {
            sum += tail / 10;
            sum += tail % 10;
        }
        else
        {
            sum += tail;
        }
        // digit ++;
        // printf("temp = %li, number = %li, tail = %li, sum = %li\n", temp, number, tail, sum);
        temp /= 100;
    }

    // 取倒数第一位
    tail = number % 10;
    sum += tail;
    // digit ++;
    // 去掉最后一位，后续计算加和按前面相似流程计算
    temp = number / 10;
    while (temp > 0)
    {
        tail = temp % 100;
        tail /= 10;
        sum += tail;
        // digit ++;
        temp /= 100;
    }

    // 保留前两位
    temp = number;
    while (!(temp > 9 && temp < 100))
    {
        temp /= 10;
        digit ++;
    }
    digit += 2;
    long digit1 = temp / 10;
    long digit2 = temp % 10;
    // 判断是不是末位为0
    int flag = 0;
    if (sum%10 == 0)
    {
        if ((digit == 13 || digit == 16) && digit1 == 4)
            flag = 1;
        else if (digit == 16 && digit1 == 5 && digit2 >= 1 && digit2 <=5)
            flag = 2;
        else if (digit == 15 && digit1 == 3 && (digit2 == 3 || digit2 == 7))
            flag = 3;
        else
            flag = 4;
    }
    else
    {
        flag = 4;
    }
    printf("digit = %li\n",digit);
    printf("sum = %li\n",sum);
    return flag;
}

File: Class6Q/sentimental-credit/credit.py
# TODO
from cs50 import get_string

def getNumber():
    while True:
        number = get_string("Number: ")
        if not number.isdigit():
            print("Please enter a number.")
        else:
            return number

def ifValid(number):
    sum = 0
    tmp = 0
    number = int(number)

    while number > 0:
        sum += number % 10
        number //= 10
        tmp = number % 10 * 2
        if tmp > 9:
            sum += tmp // 10 + tmp % 10
        else:
            sum += tmp
        number //= 10

    if sum % 10 != 0:
        print("INVALID")
        return False
    else:
        return True

def printCredit(number, isvalid):
    if isvalid:
        if len(number) == 15 and (number.startswith("37") or number .startswith("34")):
            print("AMEX")
        elif len(number) == 16 and (number.startswith("51") or number.startswith("52") or number.startswith("53") or number.startswith("54") or number.startswith("55")):
            print("MASTERCARD")
        elif (len(number) == 16 or len(number) == 13) and number.startswith("4"):
            print("VISA")
        else:
            print("INVALID")

number = getNumber()
printCredit(number, ifValid(number))
